Weblab: don't syncFiles() until bramble ready

cpirich opened the pull request:
* While investigating a new relic report, I found a possible race condition during weblab/bramble init. If `loadFileEntries()` has a `this.brambleHost`, it will try to call `syncFiles()` - which will throw if bramble is not yet ready.
* Since we have 2 separate async code paths running (bramble init, loading initial content from the files api), we now protect this window by marking `this.brambleReady` when appropriate and if our load completes before bramble is ready, we set a new `this.syncWhenBrambleReady` boolean, which defers the sync until we receive `onBrambleReady()`

islemaster commented on apps/src/weblab/WebLab.js:
```diff
@@ -444,8 +448,10 @@ WebLab.prototype.loadFileEntries = function () {
+    if (this.brambleHost && this.brambleReady) {
       this.brambleHost.syncFiles(() => {});
+    } else {
+      this.syncBrambleWhenReady = true;
```
Hmm... Having parallel asynchronous dependencies like this makes me think this is a `Promise.all` situation, though I suppose we have some hoops to jump through for that.  Could we do something like this?

```js
// @ WebLab.prototype.init where we currently call loadFileEntries
const fileEntriesLoaded = this.loadFileEntries();
const brambleReady = new Promsie((resolve) => this.reportBrambleReady = resolve);
Promise.all([fileEntriesLoaded, brambleReady]).then(() => {
  this.brambleHost.syncFiles(() => {});
});

// @ WebLab.prototype.setBrambleHost calls reportBrambleReady when Bramble is ready

// @ WebLab.prototype.loadFileEntries updated to return a Promise
```

cpirich replied:
yeah, I was thinking about something like that as well. I would have done it initially if the async operations were kicked off in a way that could be changed to return a Promise. But, what you are suggesting is a reasonable workaround to that. Will try it out...

cpirich commented:
`syncFiles()` is ok to call many times. (e.g. it happens every time you exist from the assets dialog). The real problem here relates to init - as in, when is it safe to call `syncFiles()` for the first time.

cpirich commented:
@islemaster I worked through this a little more and created some artificial delays to be sure that I understood the sequence a little better. I've now changed weblab so that it doesn't hold on to `this.brambleHost` until bramble has fired its ready event.

* Once ready, we always call `syncFiles()`
* We also call `syncFiles()` if we have a bramble host within `loadFileEntries()` - this can complete before or after we receive the 'ready' event. If before, it is safe to skip syncing (because we will once we get the 'ready'). If after, it is good to sync again.

Modified bramble host a little to ensure that we always create the root directory (if it doesn't exist) and mount the file system (if it isn't mounted) as part of the `startInitialFileSync()` - regardless of the state of the project.

islemaster commented on apps/src/weblab/brambleHost.js:
```diff
@@ -392,49 +392,62 @@ function onInspectorChanged(callback) {
   onInspectorChangedCallback_ = callback;
 }
 
+var _hasMounted = false;
```
Nit: Looks like the convention (at least within this file) is a trailing underscore for private variables, not leading.  :man_shrugging:

islemaster commented on apps/src/weblab/WebLab.js:
```diff
@@ -427,7 +449,7 @@ WebLab.prototype.onIsRunningChange = function () {
  * Load the file entry list and store it as this.fileEntries
  */
-WebLab.prototype.loadFileEntries = function () {
+WebLab.prototype.loadFileEntries = async function () {
```
Looks like there's no longer a reason to make this an `async` function, since we're not modifying any of the points that call it anymore.

islemaster commented on apps/src/weblab/WebLab.js:
```diff
@@ -379,20 +399,22 @@ WebLab.prototype.onInspectorChanged = function (inspectorOn) {
+    this.brambleHost = brambleHost;
+    brambleHost.onProjectChanged(this.onProjectChanged.bind(this));
+    brambleHost.onInspectorChanged(this.onInspectorChanged.bind(this));
+    this.brambleReady = true;
```
This flag is unused, and redundant with having `this.brambleHost` set.

islemaster commented on apps/src/weblab/brambleHost.js:
```diff
@@ -392,49 +392,62 @@ function onInspectorChanged(callback) {
+        putFilesInBramble(startSources, err => {
+          if (!err) {
+            if (forceResetToStartSources) {
+              console.log('startInitialFileSync: forceResetToStartSources - calling syncFiles()');
```
This has become an awkward structure: This `if` case only contains the `console.log` call.  I think the statement being logged still accurately represents the behavior of that case, but I can't help feeling like these conditionals should collapse now, making it obvious that we have only two paths through this logic.

```js
putFilesInBramble(startSources, (err) => {
  if (!err && !forceResetToStartSources) {
    resetBrambleChangesAndProjectVersion(currentProjectVersion);
    webLab_.registerBeforeFirstWriteHook(uploadAllFilesFromBramble);
  }
  syncFiles(wrappedCallback);
});
```

Bonus points if you can come up with a helpful name for the `!err && !forceResetToStartSources` condition. :grin:

islemaster commented on apps/src/weblab/brambleHost.js:
```diff
@@ -392,49 +392,62 @@ function onInspectorChanged(callback) {
-    });
-  }
+    } else {
+      syncFiles(wrappedCallback);
```
No need to do right now, but I think all of `startInitialFileSync` might implement much nicer using Promises or async/await.  This pair of `syncFiles(wrappedCallback)` calls smells like a 'finally' case on the end of a Promise chain that may change depending on circumstances.

cpirich replied:
yes, I agree that it could be rewritten with those tools and it would be easier to maintain
